osu.Game/Online/Chat/MessageNotifier.cs

```diff
@@ -0,0 +1,193 @@
+    {
+        [Resolved(CanBeNull = true)]
+        private NotificationOverlay notificationOverlay { get; set; }
+
+        [Resolved(CanBeNull = true)]
+        private ChatOverlay chatOverlay { get; set; }
+
+        [Resolved(CanBeNull = true)]
+        private ChannelManager channelManager { get; set; }
+
+        private Bindable<bool> notifyOnMention;
+        private Bindable<bool> notifyOnChat;
+        private Bindable<User> localUser;
+        private readonly BindableList<Channel> joinedChannels = new BindableList<Channel>();
+
+        [BackgroundDependencyLoader]
+        private void load(OsuConfigManager config, IAPIProvider api)
+        {
+            notifyOnMention = config.GetBindable<bool>(OsuSetting.ChatHighlightName);
+            notifyOnChat = config.GetBindable<bool>(OsuSetting.ChatMessageNotification);
+            localUser = api.LocalUser;
+
+            channelManager.JoinedChannels.BindTo(joinedChannels);
+
+            // Listen for new messages
+            joinedChannels.ItemsAdded += joinedChannels =>
+            {
+                foreach (var channel in joinedChannels)
+                    channel.NewMessagesArrived += newMessagesArrived;
+            };
+
+            joinedChannels.ItemsRemoved += leftChannels =>
+            {
+                foreach (var channel in leftChannels)
+                    channel.NewMessagesArrived -= newMessagesArrived;
+            };
+        }
+
+        private void newMessagesArrived(IEnumerable<Message> messages)
+        {
+            if (messages == null || !messages.Any())
+                return;
+
+            HandleMessages(messages.First().ChannelId, messages);
+        }
+
+        public void HandleMessages(long channelId, IEnumerable<Message> messages)
+        {
+            var channel = channelManager.JoinedChannels.SingleOrDefault(c => c.Id == channelId);
+
+            if (channel == null)
+            {
+                Logger.Log($"Couldn't resolve channel id {channelId}", LoggingTarget.Information);
+                return;
+            }
+
+            HandleMessages(channel, messages);
+        }
+
+        public void HandleMessages(Channel channel, IEnumerable<Message> messages)
+        {
+            // Only send notifications, if ChatOverlay and the target channel aren't visible.
+            if (chatOverlay?.IsPresent == true && channelManager.CurrentChannel.Value == channel)
+                return;
+
+            foreach (var message in messages.OrderByDescending(m => m.Id))
+            {
+                // ignore messages that already have been read
+                if (message.Id <= channel.LastReadId)
+                    return;
+
+                if (message.Sender.Id == localUser.Value.Id)
+                    continue;
+
+                // check for private messages first, if true, skip checking mentions to prevent duplicate notifications about the same message.
+                if (checkForPMs(channel, message))
```

Comment:
maybe this would be better.
```csharp
if (checkForPMs(channel, message))
    // check if the PM notification has been posted
    // to avoid potentially sending PM and mention notifications if mentioned on a PM.
    return;

Reply:
The inline comment is incorrect, as it would stop checking for all notifications altogether (note that this check is inside a `for` loop over a batch of messages). However that might actually not be entirely undesirable, if a little bit brittle.